File: AppPy/AppPy/model/WinogradScaled.py
import math
from .WinogradOriginal import WinogradOriginalImpl

def WinogradScaledImpl(matrix_A, matrix_B):
    rows_A = len(matrix_A)
    cols_A = len(matrix_A[0])
    cols_B = len(matrix_B[0])
    Result = [[0.0] * cols_B for _ in range(rows_A)]

    # Crear copias escaladas de A y B
    CopyA = [[0.0] * cols_A for _ in range(rows_A)]
    CopyB = [[0.0] * cols_B for _ in range(cols_A)]

    # Factores de escala
    a = NormInf(matrix_A, rows_A, cols_A)
    b = NormInf(matrix_B, cols_A, cols_B)
    lambda_val = math.floor(0.5 + math.log(b/a) / math.log(4))

    # Escalamiento
    MultiplyWithScalar(matrix_A, CopyA, rows_A, cols_A, 2 ** lambda_val)
    MultiplyWithScalar(matrix_B, CopyB, cols_A, cols_B, 2 ** -lambda_val)

    # Usar Winograd con matrices escaladas
    Result = WinogradOriginalImpl(CopyA, CopyB)

    return Result

# Función para calcular la norma infinita de una matriz
def NormInf(matrix, rows, cols):
    max_val = float('-inf')
    for i in range(rows):
        for j in range(cols):
            max_val = max(max_val, abs(matrix[i][j]))
    return max_val

# Función para multiplicar una matriz por un escalar
def MultiplyWithScalar(matrix, result, rows, cols, scalar):
    for i in range(rows):
        for j in range(cols):
            result[i][j] = matrix[i][j] * scalar
    return result

File: AppCs/AppCs/Algoritmos/WinogradScaled.cs
//Algoritmo basado en el libro METHODS OF MATRIX MULTIPLICATION AN OVERVIEW OF SEVERAL METHODS AND THEIR IMPLEMENTATION y apoyado en IA
using System;
using services.interfaces;

public class WinogradScaled : AlgorithmInterface
{
    /// <summary>
    /// Realiza la multiplicación de dos matrices utilizando el algoritmo de Winograd escalado.
    /// Primero, crea copias escaladas de las matrices de entrada A y B. 
    /// Luego, calcula el factor de escala lambda basado en la norma infinito de las matrices originales. 
    /// Después, escala las matrices de entrada con el factor lambda y utiliza el algoritmo de Winograd original con las matrices escaladas para obtener el resultado final. 
    /// Los parámetros de entrada son las matrices A y B, y el tamaño de las matrices junto con el tamaño del resultado (N, P, M).
    /// </summary>
    /// <param name="A">Matriz A.</param>
    /// <param name="B">Matriz B.</param>
    /// <returns>Matriz resultado de la multiplicación.</returns>
    public static long[][] Multiplication(long[][] A, long[][] B)
    {
        //Obtiene las dimensiones de las matrices
        int N = A.Length;
        int P = B[0].Length;
        int M = A[0].Length;
        //Inicializa la matriz resultado
        long[][] result = new long[N][];
        for (int j = 0; j < N; j++)
        {
            result[j] = new long[M];
        }
        int i;

        // Crear copias escaladas de A y B
        long[][] CopyA = new long[N][];
        long[][] CopyB = new long[P][];
        for (i = 0; i < N; i++)
        {
            CopyA[i] = new long[P];
        }
        for (i = 0; i < P; i++)
        {
            CopyB[i] = new long[M];
        }

        // Factor de escala
        double a = NormInf(A, N, P);
        double b = NormInf(B, P, M);
        double lambda = Math.Floor(0.5 + Math.Log(b / a) / Math.Log(4));

        // Escalar
        Util.MultiplyWithScalar(A, CopyA, N, P, Math.Pow(2, lambda));
        Util.MultiplyWithScalar(B, CopyB, P, M, Math.Pow(2, -lambda));

        // Utilizar Winograd con las matrices escaladas
        result = Winograd.Original(CopyA, CopyB);

        return result;
    }

    /// <summary>
    /// Calcula la norma infinito de una matriz.
    /// </summary>
    /// <param name="matrix">La matriz.</param>
    /// <param name="rows">El número de filas de la matriz.</param>
    /// <param name="cols">El número de columnas de la matriz.</param>
    /// <returns>La norma infinito de la matriz.</returns>
    private static double NormInf(long[][] matrix, int rows, int cols)
    {
        double maxNorm = 0;
        for (int i = 0; i < rows; i++)
        {
            double rowSum = 0;
            for (int j = 0; j < cols; j++)
            {
                rowSum += Math.Abs(matrix[i][j]);
            }
            maxNorm = Math.Max(maxNorm, rowSum);
        }
        return maxNorm;
    }

    public long[][] MultiplyMatrices(long[][] matrix1, long[][] matrix2)
    {
        return Multiplication(matrix1,matrix2);
    }
}
